src/data_access/api/SpotifyAPI.java

```diff
@@ -0,0 +1,109 @@
+package data_access.api;
+
+import entity.CommonSongFactory;
+import entity.Song;
+import entity.SongFactory;
+import io.github.cdimascio.dotenv.Dotenv;
+import okhttp3.*;
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Random;
+
+public class SpotifyAPI implements SongAPI{
+    private final String CLIENT_ID;
+    private final String CLIENT_SECRET;
+    private final OkHttpClient client;
+    private String authToken = null;
+    private final SongFactory songFactory;
+
+    public SpotifyAPI(SongFactory songFactory){
+        Dotenv dotenv = Dotenv.configure().filename("api.env").load();
+        this.CLIENT_ID = dotenv.get("CLIENT_ID");
+        this.CLIENT_SECRET = dotenv.get("CLIENT_SECRET");
+        this.client = new OkHttpClient().newBuilder().build();
+        this.songFactory = songFactory;
+    }
+
+    @Override
+    public Song getRandomSongFromGenre(String genre) {
+        JSONObject responseBody = retrieveSongData(genre);
+        JSONArray songsArray = responseBody.getJSONObject("tracks").getJSONArray("items");
+        JSONObject item;
+        Song song = null;
+        int i = 0;
+        do{
+            item = (JSONObject) songsArray.get(i);
+            if (item.get("preview_url") instanceof String){
+                JSONObject albumArtistInfo = (JSONObject) item.getJSONObject("album").getJSONArray("artists").get(0);
+                String songName = item.getString("name");
+                String audio = item.getString("preview_url");
+                String artistName = albumArtistInfo.getString("name");
+                song = songFactory.create(songName, artistName, audio);
+            }
+            i++;
+        }while (!(item.get("preview_url") instanceof String)  || item.getInt("popularity") < 83);
+
+        if(song == null){
+            return getRandomSongFromGenre(genre);
+        }else {
+            return song;
+        }
+    }
+    private JSONObject retrieveSongData(String genre)  {
+        String url = getRandomEndpoint(genre);
+        JSONObject response;
+        try {
+            Request request = new Request.Builder()
+                    .url(url)
+                    .addHeader("Authorization", "Bearer " + authToken)
+                    .addHeader("Content-Type", "application/json")
+                    .build();
+
+            Response responseRequest = client.newCall(request).execute();
```

Comment:
```suggestion
        Request request = new Request.Builder()
                .url(url)
                .addHeader("Authorization", "Bearer " + authToken)
                .addHeader("Content-Type", "application/json")
                .build();
        try (Response responseRequest = client.newCall(request).execute()) {
```
I had a couple of error messages show up (albeit inconsistently) regarding resource leaks with SpotifyAPI. I believe this has to do with not properly closing the client connection. One way to fix this is to use a try-with statement as shown above to automatically close the connection after the block finishes.